doc_generator/generate.py

```diff
@@ -1,178 +1,204 @@
-    f.write("        Manufacturing: 3D Printing\n")
-    f.write("        Material: PLA or PETG\n")
-    #f.write("        Preview: [preview](models/beam.stl){previewpage}\n")
-
-with open(outputdir_gitbuilding+'DeviceParts.yaml', 'w') as f:
-  for (part, device) in listOfTrays:
-    f.write("%s:\n" % (device['ID']))
-    f.write("    Name: %s\n" % (device['Name']))
-    f.write("    Specs:\n")
-    for k in device.keys():
-      f.write("        %s: %s\n" % (k, device[k]))
-
-with open(outputdir_gitbuilding+'components.md', 'w') as f:
-    f.write("# Installing the components in trays\n\n")
-    f.write("{{BOM}}\n")
-    f.write("![](svg/baseplate_beams_topplate.svg)\n\n")
-    f.write("For all of your components:\n\n")
-    f.write("* find the corresponding 3d printed tray\n")
-    f.write("* install the tray on the rack\n")
-    f.write("* mount the device on the tray\n\n")
-    f.write("Here are the devices and the trays\n\n")
-    for (part, device) in listOfTrays:
-      f.write("* [%s](DeviceParts.yaml#%s){Qty: 1}\n" % (device['Name'], device['ID']))
-      stlfile = F"tray_{device['ID']}.stl"
-      f.write("* %s ([preview](models/%s){previewpage}, [download](models/%s))\n" % (stlfile, stlfile, stlfile))
-    f.write("![](svg/trays.svg)")
-
-#for debugging
-#show_object(beam)
+from pathlib import Path
+import shutil
 
+import cadquery as cq
 
+from utility.export import export_svg
+from utility.export import export_stl
+from utility.export import export_step
+
+import mechanical.components.cadquery.rack_leg as rack_leg
+import mechanical.components.cadquery.nimble_top as nimble_top
+import mechanical.components.cadquery.nimble_end_plate as nimble_end_plate
+import mechanical.components.cadquery.nimble_tray as nimble_tray
+
+
+
+def generate_docs(config, config_hash, force_rebuild=True):
+    """
+    generate all models and update documentation
+    """
+
+    # parameters
+
+    #TODO: get from input
+    #selected_devices_ids = config.devices
+    selected_devices_ids = ['NUC10i5FNH','RPi4']
+
+    # get path of this python file
+    module_path = Path(__file__).resolve().parent
+
+    build_dir = module_path  / "build" / config_hash
+    build_dir = build_dir.resolve()
+    build_dir_identification_file = "NIMBLE_BUILD"
+
+    if os.path.exists(build_dir):
+        if force_rebuild:
+            # for safety, only delete if build dir is identified as such
+            if not os.path.exists(build_dir / build_dir_identification_file):
+                print("build dir exists but is not identified as such, refusing to delete")
+                raise Exception("build dir exists but is not identified as such, refusing to delete")
+            shutil.rmtree(build_dir)
+        else:
+            print("build dir already exists, skipping")
+            return
+        
+    json_filename = module_path / "devices.json"
+    outputdir_gitbuilding = build_dir / "gitbuilding/"
```

Comment:
The end of this line has a forward slash after gitbuilding. Does that break cross-platform compatibility?

Reply:
Forward slashes work nicely on Windows too (when using standard python library methods), but as this slash is unnecessary, I'll remove it anyway